### PR TITLE
Proto changes for pattern template extensions

### PR DESCRIPTION
**Pull request by silverstar194**
Signed-off-by: silverstar195 <seanmaloney@google.com>

Commit Message:
Proto changes for pattern template extensions

Additional Description:
This PR is the first of a series to implement pattern matching and rewrite functionality for Envoy.

These specific proto changes:
1. Add the foundation for two extension configs
2. Change the API to expose the extension configs to end users

Additional follow up PR to implement extension can be found here and is a WIP:
[Add match and rewrite extension implementations ](https://github.com/envoyproxy/envoy/pull/22214)
[Use extensions in data and control plane](https://github.com/envoyproxy/envoy/pull/22207)

Risk Level:
Low

Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (phlax) [inline api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:26]:
```suggestion
// or ``path_template_rewrite`` may be specified.
```

**Review by phlax — DISMISSED** [review]
@silverstar194 thanks for working on this

please ensure all literals use double backticks for correct rst rendering

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (edin44) [inline api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:61]: based

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (edin44) [inline api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:72]: "is the index into" or "indexes into" instead of "index into"?

**Review by edin44 — DISMISSED** [review]

File: source/common/router/config_impl.cc
Comment (stevenzzzz) [inline source/common/router/config_impl.cc:1385]: the pattern is raise a unimplemented exception.

**Review by stevenzzzz — DISMISSED** [review]
LGTM. with some minor nits.

File: source/common/router/config_impl.cc
Comment (silverstar194) [reply source/common/router/config_impl.cc:1385]: changed

File: api/envoy/config/route/v3/route_components.proto
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:1059]: changed to match others

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:70]: Editor default. Can change if needed. Couldn't find Envoy style guide. What is the recommend length?

Comment (phlax) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:70]: i dont think we have one for proto/rst - but i would say an ideal of 100 if not strictly enforced

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:58]: No reason now that it is in config proto. Will remove

Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:58]: removed here as well

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:56]: Would prefer to keep all the proto changes together

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:62]: Will remove from this PR and move closer to extension

File: source/common/router/config_impl.cc
Comment (silverstar194) [inline source/common/router/config_impl.cc:1385]: Removed unused/to be implemented lines to meet code coverage

**Comment by htuch** [comment]
API LGTM but needs to be properly hooked up into the extension category and instance annotations and metadata.

File: api/envoy/config/route/v3/route_components.proto
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:575]: Will add as a TODO. Can't add yet as it isn't present and will fail validation.

File: api/envoy/config/route/v3/route_components.proto
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:1080]: Same as above

**Review by stevenzzzz — DISMISSED** [review]
LGTM again.

**Comment by silverstar194** [comment]
@htuch Ready to merge when you approve. Just blocked on API approver

**Review by phlax — CHANGES_REQUESTED** [review]
@silverstar194 ive left a couple of docs nits inline - i didnt review the non-implemented docs as they are not currently rendered

File: api/envoy/config/route/v3/route_components.proto
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:1024]: What part would be a link?

Comment (phlax) [reply api/envoy/config/route/v3/route_components.proto:1024]: `prefix_rewrite`

Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto]: changed

File: api/envoy/config/route/v3/route_components.proto
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto]: changed

File: api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto
Comment (htuch) [inline api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto:3]: Sorry, one last nit here. You've created two new extension categories, but are placing two concrete extension instances for each under a common tree. I would instead organize this as two trees:
* `envoy.extensions.path.match`
* `envoy.extensions.path.rewrite`

Then you have the two examples/default instances underneath these as:
* `envoy.extensions.path.match.pattern_template.v3.`
* `envoy.extensions.path.rewrite.pattern_template.v3`

I.e. you have two generic path match/rewrite extension categories, and the "pattern template" syntax in each case is just an instantiation of the category.

Comment (silverstar194) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: Sounds good. Switched over

**Comment by silverstar194** [comment]
@lizan Feel free to merge. Don't have permissions on my end

File: api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto
Comment (htuch) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: Sort of - you should have pattern_template as a path component as suggested ^^ to make clear that this is one example of a "path match" or a "path rewrite".

File: api/envoy/config/route/v3/route_components.proto
Comment (lizan) [inline api/envoy/config/route/v3/route_components.proto:575]: mark these not-implemented-hide?

**Review by lizan — COMMENTED** [review]
I wasn't meant final approval, if this merges first then let's sort out the comment in proto.

**Review by phlax — DISMISSED** [review]

**Comment by phlax** [comment]
thanks for fixing doc links @silverstar194

File: api/envoy/config/route/v3/route_components.proto
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:576]: Done.

File: api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: I believe this is switched. Comment is on the old file path.

Comment (silverstar194) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: Ok, think I got it right now. Added the additional directory pattern_template

**Review by lizan — APPROVED** [review]

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
